Zed crashes repeatedly when interacting with the Git hunk view

### Summary

<!-- Please insert a one line summary of the issue below -->
Zed crashes when selecting deleted code in Git hunk view  or scrolling through it. I have been able to reproduce the crash only with a single Git hunk view in my code.

<!-- Include all steps necessary to reproduce from a clean Zed installation. Be verbose -->
Steps to trigger the problem:
1. Open a deleted code git hunk view 
2. Just select all the code

Actual Behavior:
Zed crash

Expected Behavior:
Been able to select the code

### Zed Version and System Specs

Zed: v0.172.1 (Zed Preview)
OS: macOS 15.3.0
Memory: 48 GiB
Architecture: aarch64

### If applicable, attach your `~/Library/Logs/Zed/Zed.log` file to this issue.

<details><summary>Zed.log</summary>

<!-- Click below this line and paste or drag-and-drop your log-->
```
2025-01-30T22:51:15.00336+01:00 [ERROR] {
  "thread": "main",
  "payload": "offset 991 is out of range, max allowed is 937",
  "location_data": {
    "file": "crates/text/src/text.rs",
    "line": 2938
  },
  "backtrace": [
    "backtrace::capture::Backtrace::create::hc4c89705da5b1737+10162008",
    "backtrace::capture::Backtrace::new::hf7d621b6b1194da3+10161820",
    "zed::reliability::init_panic_hook::{{closure}}::h1ee27e16989fe402+82778372",
    "std::panicking::rust_panic_with_hook::h09e8a656f11e82b2+63673596",
    "std::panicking::begin_panic_handler::{{closure}}::h1230eb3cc91b241c+63670692",
    "std::sys::backtrace::__rust_end_short_backtrace::hc3491307aceda2c2+63662796",
    "_rust_begin_unwind+63669880",
    "core::panicking::panic_fmt::ha4b80a05b9fff47a+91299092",
    "multi_buffer::MultiBufferSnapshot::lift_buffer_metadata::{{closure}}::hf73b9c40833e47ee+19417684",
    "<editor::element::EditorElement as gpui::element::Element>::prepaint::{{closure}}::{{closure}}::{{closure}}::hbb3961832b9bd884+22382336",
    "<editor::element::EditorElement as gpui::element::Element>::prepaint::{{closure}}::hf6ce17ebcba74d23+22357844",
    "<editor::element::EditorElement as gpui::element::Element>::prepaint::hed896feb6fc7beee+22356864",
    "<gpui::element::Drawable<E> as gpui::element::ElementObject>::prepaint::h4e670dfa1877b40e+21389648",
    "gpui::element::AnyElement::prepaint::hd85bb4ef651de54a+32500720",
    "<gpui::view::AnyView as gpui::element::Element>::prepaint::hec3b73551f331600+32939508",
    "<gpui::element::Drawable<E> as gpui::element::ElementObject>::prepaint::hc2eefea6952ab01d+78586228",
    "gpui::element::AnyElement::prepaint::hd85bb4ef651de54a+32500720",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}::{{closure}}::{{closure}}::h20b0e5a99413158a+32535140",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}::{{closure}}::hef00bc3261b51497+32534528",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}::h3a17c496bf78c752+32526544",
    "<gpui::elements::div::Div as gpui::element::Element>::prepaint::h3f5b79fd07e068b9+32510900",
    "<gpui::element::Drawable<E> as gpui::element::ElementObject>::prepaint::h91748ce6feab8aeb+78574304",
    "gpui::element::AnyElement::prepaint::hd85bb4ef651de54a+32500720",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}::{{closure}}::{{closure}}::h20b0e5a99413158a+32535140",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}::{{closure}}::hef00bc3261b51497+32534528",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}::h3a17c496bf78c752+32526544",
    "<gpui::elements::div::Div as gpui::element::Element>::prepaint::h3f5b79fd07e068b9+32510900",
    "<gpui::element::Drawable<E> as gpui::element::ElementObject>::prepaint::h91748ce6feab8aeb+78574304",
    "gpui::element::AnyElement::prepaint::hd85bb4ef651de54a+32500720",
    "<gpui::view::AnyView as gpui::element::Element>::prepaint::{{closure}}::hd435952f47b9ca3d+32942768",
    "<gpui::view::AnyView as gpui::element::Element>::prepaint::hec3b73551f331600+32940640",
    "<gpui::element::Drawable<E> as gpui::element::ElementObject>::prepaint::hc2eefea6952ab01d+78586228",
    "gpui::element::AnyElement::prepaint::hd85bb4ef651de54a+32500720",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}::{{closure}}::{{closure}}::h20b0e5a99413158a+32535140",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}::{{closure}}::hef00bc3261b51497+32534320",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}::h3a17c496bf78c752+32526544",
    "<gpui::elements::div::Div as gpui::element::Element>::prepaint::h3f5b79fd07e068b9+32510900",
    "<gpui::element::Drawable<E> as gpui::element::ElementObject>::prepaint::h91748ce6feab8aeb+78574304",
    "gpui::element::AnyElement::prepaint::hd85bb4ef651de54a+32500720",
    "<workspace::pane_group::element::PaneAxisElement as gpui::element::Element>::prepaint::h29ac2fcd85949e16+79037512",
    "<gpui::element::Drawable<E> as gpui::element::ElementObject>::prepaint::h19e0cd1dcf5a189f+78554612",
    "gpui::element::AnyElement::prepaint::hd85bb4ef651de54a+32500720",
    "<gpui::element::Drawable<E> as gpui::element::ElementObject>::prepaint::h30fb0c5932287973+78560216",
    "gpui::element::AnyElement::prepaint::hd85bb4ef651de54a+32500720",
    "<gpui::element::Drawable<E> as gpui::element::ElementObject>::prepaint::h30fb0c5932287973+78560216",
    "gpui::element::AnyElement::prepaint::hd85bb4ef651de54a+32500720",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}::{{closure}}::{{closure}}::h20b0e5a99413158a+32535140",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}::{{closure}}::hef00bc3261b51497+32534320",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}::h3a17c496bf78c752+32526544",
    "<gpui::elements::div::Div as gpui::element::Element>::prepaint::h3f5b79fd07e068b9+32510900",
    "<gpui::element::Drawable<E> as gpui::element::ElementObject>::prepaint::h91748ce6feab8aeb+78574304",
    "gpui::element::AnyElement::prepaint::hd85bb4ef651de54a+32500720",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}::{{closure}}::{{closure}}::h20b0e5a99413158a+32535140",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}::{{closure}}::hef00bc3261b51497+32534528",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}::h3a17c496bf78c752+32526544",
    "<gpui::elements::div::Div as gpui::element::Element>::prepaint::h3f5b79fd07e068b9+32510900",
    "<gpui::element::Drawable<E> as gpui::element::ElementObject>::prepaint::h91748ce6feab8aeb+78574304",
    "gpui::element::AnyElement::prepaint::hd85bb4ef651de54a+32500720",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}::{{closure}}::{{closure}}::h20b0e5a99413158a+32535140",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}::{{closure}}::hef00bc3261b51497+32534320",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}::h3a17c496bf78c752+32526544",
    "<gpui::elements::div::Div as gpui::element::Element>::prepaint::h3f5b79fd07e068b9+32510900",
    "<gpui::element::Drawable<E> as gpui::element::ElementObject>::prepaint::h91748ce6feab8aeb+78574304",
    "gpui::element::AnyElement::prepaint::hd85bb4ef651de54a+32500720",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}::{{closure}}::{{closure}}::h20b0e5a99413158a+32535140",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}::{{closure}}::hef00bc3261b51497+32534528",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}::h3a17c496bf78c752+32526544",
    "<gpui::elements::div::Div as gpui::element::Element>::prepaint::h3f5b79fd07e068b9+32511512",
    "<gpui::element::Drawable<E> as gpui::element::ElementObject>::prepaint::h91748ce6feab8aeb+78574304",
    "gpui::element::AnyElement::prepaint::hd85bb4ef651de54a+32500720",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}::{{closure}}::{{closure}}::h20b0e5a99413158a+32535140",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}::{{closure}}::hef00bc3261b51497+32534320",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}::h3a17c496bf78c752+32526544",
    "<gpui::elements::div::Div as gpui::element::Element>::prepaint::h3f5b79fd07e068b9+32510900",
    "<gpui::element::Drawable<E> as gpui::element::ElementObject>::prepaint::h91748ce6feab8aeb+78574304",
    "gpui::element::AnyElement::prepaint::hd85bb4ef651de54a+32500720",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}::{{closure}}::{{closure}}::h20b0e5a99413158a+32535140",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}::{{closure}}::hef00bc3261b51497+32534528",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}::h3a17c496bf78c752+32526820",
    "<gpui::elements::div::Div as gpui::element::Element>::prepaint::h3f5b79fd07e068b9+32510900",
    "<gpui::element::Drawable<E> as gpui::element::ElementObject>::prepaint::h91748ce6feab8aeb+78574304",
    "gpui::element::AnyElement::prepaint::hd85bb4ef651de54a+32500720",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}::{{closure}}::{{closure}}::h20b0e5a99413158a+32535140",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}::{{closure}}::hef00bc3261b51497+32534320",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}::h3a17c496bf78c752+32526544",
    "<gpui::elements::div::Div as gpui::element::Element>::prepaint::h3f5b79fd07e068b9+32510900",
    "<gpui::element::Drawable<E> as gpui::element::ElementObject>::prepaint::h91748ce6feab8aeb+78574304",
    "gpui::element::AnyElement::prepaint::hd85bb4ef651de54a+32500720",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}::{{closure}}::{{closure}}::h20b0e5a99413158a+32535140",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}::{{closure}}::hef00bc3261b51497+32534320",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}::h3a17c496bf78c752+32526544",
    "<gpui::elements::div::Div as gpui::element::Element>::prepaint::h3f5b79fd07e068b9+32511512",
    "<gpui::element::Drawable<E> as gpui::element::ElementObject>::prepaint::h91748ce6feab8aeb+78574304",
    "gpui::element::AnyElement::prepaint::hd85bb4ef651de54a+32500720",
    "<gpui::view::AnyView as gpui::element::Element>::prepaint::hec3b73551f331600+32939508",
    "<gpui::element::Drawable<E> as gpui::element::ElementObject>::prepaint::hefdd159a3cc7288a+32493064",
    "gpui::element::AnyElement::prepaint::hd85bb4ef651de54a+32500720",
    "gpui::window::Window::draw::hefb802f4ded83242+33084976",
    "gpui::window::Window::new::{{closure}}::{{closure}}::h388d519de5e969fa+32976908",
    "gpui::window::Window::new::{{closure}}::hd1fcb123eb11d22d+32971592",
    "gpui::platform::mac::window::step::hd3f5b22f168c2109+32845512",
    "<gpui::platform::mac::platform::MacPlatform as gpui::platform::Platform>::run::he7d5379a29d93919+32765352",
    "gpui::app::Application::run::hb3fdb527680bc694+80793024",
    "zed::main::h6fcb4bb0c3c6ad04+83434808",
    "std::sys::backtrace::__rust_begin_short_backtrace::h0f15a40bdb03c005+81748160",
    "std::rt::lang_start::{{closure}}::h3a59a78b02624e1f+82815476",
    "std::rt::lang_start_internal::hdd117cb81a316264+63604276",
    "_main+83447112"
  ],
  "app_version": "0.172.1",
  "release_channel": "preview",
  "target": "aarch64-apple-darwin",
  "os_name": "macOS",
  "os_version": "15.3.0",
  "architecture": "aarch64",
  "panicked_on": 1738273875003,
  "system_id": "7b3bb1c7-c398-42be-b06a-3f27cebc4022",
  "installation_id": "cdce0e20-e65a-4cc1-8abc-d81bbfb63992",
  "session_id": "7c39e21d-0448-45ea-9b3c-2b8925c6318e"
}
```
<!-- Click above this line and paste or drag-and-drop your log--></details>

--- a/crates/editor/src/items.rs
+++ b/crates/editor/src/items.rs
@@ -1040,10 +1040,10 @@ impl SerializableItem for Editor {
             } => window.spawn(cx, |mut cx| {
                 let project = project.clone();
                 async move {
-                    let language = if let Some(language_name) = language {
-                        let language_registry =
-                            project.update(&mut cx, |project, _| project.languages().clone())?;
+                    let language_registry =
+                        project.update(&mut cx, |project, _| project.languages().clone())?;
 
+                    let language = if let Some(language_name) = language {
                         // We don't fail here, because we'd rather not set the language if the name changed
                         // than fail to restore the buffer.
                         language_registry
@@ -1061,6 +1061,7 @@ impl SerializableItem for Editor {
 
                     // Then set the text so that the dirty bit is set correctly
                     buffer.update(&mut cx, |buffer, cx| {
+                        buffer.set_language_registry(language_registry);
                         if let Some(language) = language {
                             buffer.set_language(Some(language), cx);
                         }

--- a/crates/multi_buffer/src/multi_buffer.rs
+++ b/crates/multi_buffer/src/multi_buffer.rs
@@ -3484,7 +3484,8 @@ impl MultiBufferSnapshot {
                 let region = cursor.region()?;
                 let buffer_start = if region.is_main_buffer {
                     let start_overshoot = range.start.saturating_sub(region.range.start.key);
-                    region.buffer_range.start.key + start_overshoot
+                    (region.buffer_range.start.key + start_overshoot)
+                        .min(region.buffer_range.end.key)
                 } else {
                     cursor.main_buffer_position()?.key
                 };

--- a/crates/project/src/buffer_store.rs
+++ b/crates/project/src/buffer_store.rs
@@ -2223,7 +2223,9 @@ impl BufferChangeSet {
                         let base_text = cx.background_executor().spawn(snapshot).await;
                         this.update(&mut cx, |this, cx| {
                             this.base_text = Some(base_text);
-                            cx.notify();
+                            cx.emit(BufferChangeSetEvent::DiffChanged {
+                                changed_range: text::Anchor::MIN..text::Anchor::MAX,
+                            });
                         })
                     }));
                 }

